compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -820,6 +820,9 @@ impl ObjectSafetyViolation {
             ) => {
                 format!("method `{name}` references an `impl Trait` type in its return type").into()
             }
+            ObjectSafetyViolation::Method(name, MethodViolationCode::AlreadyErrored, _) => {
```

Comment:
Hm. It seems weird that we report this as an object safety violation. I assume it also causes downstream errors like forcing implementijgn type not to be able to be coerced to `&dyn Trait`.

I'm somewhat surprised that we don't have any tests affected by this. What does it do when you have something like:

```
trait Foo {
  fn err(&self) -> MissingType;
}

impl Foo for i32 {
  fn err(&self) -> MissingType { 0 }
}

fn coerce(x: &i32) -> &dyn Foo { x }
```

Reply:
```
error[E0038]: the trait `Foo` cannot be made into an object
  --> $DIR/erroneous_signature.rs:11:24
   |
LL | fn coerce(x: &i32) -> &dyn Foo {
   |                        ^^^^^^^ `Foo` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
  --> $DIR/erroneous_signature.rs:2:8
   |
LL | trait Foo {
   |       --- this trait cannot be made into an object...
LL |     fn err(&self) -> MissingType;
   |        ^^^ ...because method `err` has errors in its signature
   = help: consider moving `err` to another trait
   = help: only type `i32` implements the trait, consider using it directly instead
```

I can also make the object safety queries return `Result<_, ErrorGuaranteed>` to avoid the bogus messages.

Reply:
Yeah, that's what I expected it to say.

> I can also make the object safety queries return Result<_, ErrorGuaranteed> to avoid the bogus messages.

I don't think we really should kill all the other object safety violations though. Some of them will be valid.

Unless we have some other reason to suppress object safety violations here, I kind of just prefer the approach in the other PR, since it's more limited.

Reply:
er, not other PR, the previous approach. I thought you opened this as a separate PR 😆

Reply:
Well... the previous approach was also wrong

```
LL | trait Foo {
   |       --- this trait cannot be made into an object...
LL |     async fn foo(self: &dyn Foo) {
   |              ^^^ ...because method `foo` is `async`
```

Reply:
ah jup 🤦